### PR TITLE
Expose SemanticsUpdate and SemanticsUpdateBuilder

### PR DESCRIPTION
**Pull request by abarth**
This patch prepares us to switch sending semantics information over mojom to
using dart:ui directly. Currently the recorded information is dropped on the
floor, but that will improve in future patches.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

File: lib/ui/semantics.dart [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+/// [Window.updateSemantics] to update the semantics conveyed to the user.
+class SemanticsUpdateBuilder extends NativeFieldWrapperClass2 {
+  /// Creates an empty [SemanticsUpdateBuilder] object.
+  SceneBuilder() { _constructor(); }
+  void _constructor() native "SemanticsUpdateBuilder_constructor";
+
+  /// Update the information associated with the node with the given `id`.
+  ///
+  /// The semantics nodes form a tree, with the root of the tree always having
+  /// an id of zero. The `children` are the ids of the nodes that are immeidate
```
Comment (Hixie) [inline lib/ui/semantics.dart]: immediate

File: lib/ui/semantics.dart [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  /// children of this node. The system retains the nodes that are currently
+  /// reachable from the root. A given update need not contain information for
+  /// nodes that do not change in the update. If a node is not reachable from
+  /// the root after an update, the node will be discarded from the tree.
+  ///
+  /// The `flags` are a bitfield of [SemanticsFlags] that apply to this node.
+  ///
+  /// The `actions` are a bit field of [SemanticsActions] that can be undertaken
```
Comment (Hixie) [inline lib/ui/semantics.dart]: decide if it's `bit field` or `bitfield`

File: lib/ui/semantics.dart [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  /// the possible [SemanticsActions]. Because the semantics tree is maintained
+  /// asynchronously, the [Window.onSemanticsAction] callback might be called
+  /// with an action that is no longer possible.
+  ///
+  /// The `label` is a string that describes this node.
+  ///
+  /// The `rect` is the region occupied by this node in its own coordinate
+  /// system.
+  ///
+  /// The `transform` is a matrix that maps this nodes coodinate system into its
```
Comment (Hixie) [inline lib/ui/semantics.dart]: node's

File: lib/ui/semantics.dart [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    int id,
+    int flags,
+    int actions,
+    Rect rect,
+    String label,
+    Float64List transform,
+    Int32List children
+  }) {
+    if (transform.length != 16)
+      throw new ArgumentError("[transform] must have 16 entries.");
```
Comment (Hixie) [inline lib/ui/semantics.dart]: no square brackets

File: lib/ui/semantics.dart [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+  ///
+  /// Calling the SemanticsUpdate constructor directly will not create a useable
+  /// object. To create a SemanticsUpdate object, use a [SemanticsUpdateBuilder].
+  SemanticsUpdate(); // (this constructor is here just so we can document it)
+
+  /// Releases the resources used by this semantics update.
+  ///
+  /// After calling this function, the semantics update is cannot be used
+  /// further.
+  void dispose() native "Scene_dispose";
```
Comment (Hixie) [inline lib/ui/semantics.dart]: Scene_dispose? scary copy pasta!

File: lib/ui/window.dart [hunk 6]
```diff
@@ -179,6 +185,30 @@ class Window {
...
+  /// performed.
+  ///
+  /// This callback is used when the user expresses the action they wish to
+  /// perform based on the semantics supplied by [updateSemantics].
+  SemanticsActionCallback onSemanticsAction;
+
+  /// Change the retained semantics data about this window.
+  ///
+  /// If [semanticsEnabled] is true, the user has requested that this funciton
+  /// be called whenever the semantic content of this window changes.
```
Comment (Hixie) [inline lib/ui/window.dart]: does this dispose the update for you or do you then have to update it?

document the answer

**Comment by Hixie** [comment]
LGTM, but travis is angry.

File: lib/ui/semantics.dart (hunk 2 above)
Comment (abarth) [inline lib/ui/semantics.dart]: Wikipedia says `bit field`

File: lib/ui/semantics.dart (hunk 4 above)
Comment (abarth) [inline lib/ui/semantics.dart]: We use `[` consistently in all these ArgumentErrors.  I'll leave this for now and do a separate patch to remove them all.

**Pull request merged**